dipy/nn/deepn4.py

```diff
@@ -0,0 +1,332 @@
+                     strides=1, padding='valid')(x)
+
+    # Last layer without relu
+    out = Conv3DTranspose(1, kernel_size=1,
+                          strides=1, padding='valid')(x)
+
+    return Model(inputs, out)
+
+
+class DeepN4:
+    """
+    This class is intended for the DeepN4 model.
+    """
+
+    @doctest_skip_parser
+    def __init__(self, verbose=False):
+        r"""
+
+        To obtain the pre-trained model, use::
+        >>> deepn4_model = DeepN4() # skip if not have_tf
+        >>> fetch_model_weights_path = get_fnames('model_weights')
+        >>> skip fetch_model_weights_path not have_tf
+        >>> deepn4_model.load_model_weights(fetch_model_weights_path)
+        >>> skip deepn4_model not have_tf
+
+        This model is designed to take as input file T1 signal and predict
+        bias field. Effectively, this model is mimicking bias correction.
+
+        Parameters
+        ----------
+        verbose : bool (optional)
+            Whether to show information about the processing.
+            Default: False
+
+        References
+        ----------
+        Kanakaraj, P., Yao, T., Cai, L. Y., Lee, H. H., Newlin, N. R.,
+        Kim, M. E., & Moyer, D. (2023). DeepN4: Learning N4ITK Bias Field
+        Correction for T1-weighted Images.
+        """
+
+        if not have_tf:
+            raise tf()
+        if not have_tfa:
+            raise tfa()
+
+        log_level = 'INFO' if verbose else 'CRITICAL'
+        set_logger_level(log_level, logger)
+
+        # Synb0 network load
+
+        self.model = UNet3D(input_shape=(128, 128, 128, 1))
+
+    def fetch_default_weights(self):
+        r"""
+        Load the model pre-training weights to use for the fitting.
+        Will not work if the declared SH_ORDER does not match the weights
+        expected input.
+        """
+        fetch_model_weights_path = get_fnames('deepn4_default_weights')
+        self.load_model_weights(fetch_model_weights_path)
+
+    def load_model_weights(self, weights_path):
+        r"""
+        Load the custom pre-training weights to use for the fitting.
+        Will not work if the declared SH_ORDER does not match the weights
+        expected input.
+
+        The weights for a sh_order of 8 can be obtained via the function:
+            get_fnames('deepn4_default_weights').
+
+        Parameters
+        ----------
+        weights_path : str
+            Path to the file containing the weights (hdf5, saved by tensorflow)
+        """
+        try:
+            self.model.load_weights(weights_path)
+        except ValueError:
+            raise ValueError('Expected input for the provided model weights \
+                             do not match the declared model')
+
+    def __predict(self, x_test):
+        r"""
+        Internal prediction function
+        Predict bias field from input T1 signal
+
+        Parameters
+        ----------
+        x_test : np.ndarray (1, 1, 128, 128, 128)
```

Comment:
Channel is last dimension.

Reply:
changed it to `x_test : np.ndarray (128, 128, 128, 1)`